### PR TITLE
Mypy for intervals

### PR DESCRIPTION
**Pull request by AdrianSosic**
Enables `mypy` for the `Interval` class, which are used by targets (done in #54) and parameters (done in #28). Along the way, adds missing type hints and cleans up the class properties.

In particular, replaces the somewhat misleading `is_finite` and `is_bounded` methods with the common and easy-to-understand mathematical definitions `is_open`, `is_half_open` and `is_closed`, ensuring backward compatibility via deprecation mechanisms.

@AVHopp , @Scienfitz: Note that the PR builds upon #54, in order to configure the right diff view.

### CODE REVIEW

File: baybe/utils/interval.py [hunk 1]
```diff
@@ -41,40 +45,66 @@ class Interval:
...
+        if upper <= self.lower:
             raise ValueError(
-                f"The upper interval bound (provided value: {value}) must be larger "
+                f"The upper interval bound (provided value: {upper}) must be larger "
                 f"than the lower bound (provided value: {self.lower})."
             )
 
     @property
-    def is_finite(self):
+    def is_closed(self) -> bool:
```
Comment (AVHopp) [inline baybe/utils/interval.py:61]: This naming could yield confusion: Note that a closed interval is typically interpreted as an interval containing both end points, while an open interval is one that does NOT contain either of the endpoints. For example, the interval (0,1) not containing the numbers 0 and 1 is an open but finite interval. In this terminology, this would however be a *closed* interval which is not the correct term here. I guess what you mean here is better called bounded?

File: baybe/utils/interval.py [hunk 2]
```diff
@@ -41,40 +45,66 @@ class Interval:
...
+        """Check whether the interval is closed."""
+        return np.isfinite(self.lower) and np.isfinite(self.upper)
+
+    @property
+    def is_half_open(self) -> bool:
+        """Check whether the interval is half-open."""
+        return np.isfinite(self.lower) ^ np.isfinite(self.upper)
+
+    @property
+    def is_open(self) -> bool:
```
Comment (AVHopp) [inline baybe/utils/interval.py:71]: See my comment at `is_closed` - using open here is *not* correct from a mathematical point.

File: baybe/utils/interval.py (hunk 1 above)
Comment (AVHopp) [reply baybe/utils/interval.py:61]: Unfortunately, we seem to have used this term previously, so I guess this might yield problems resp. force us to already get rid of the backwards compatibility?

File: baybe/utils/interval.py [hunk 3]
```diff
@@ -90,15 +120,15 @@ def _(cls, bounds: Iterable):
...
 
-    def to_tuple(self):
-        """Transfor the interval to a tuple."""
+    def to_tuple(self) -> Tuple[float, float]:
+        """Transform the interval to a tuple."""
         return self.lower, self.upper
 
-    def to_ndarray(self):
+    def to_ndarray(self) -> np.ndarray:
         """Transform the interval to a ndarray."""
```
Comment (AVHopp) [inline baybe/utils/interval.py:128]: Please add backticks around the `ndarray`.

File: mypy.ini [hunk 4]
```diff
@@ -8,7 +8,14 @@ exclude = (?x)(
           | baybe/searchspace
           | baybe/strategies
           | baybe/surrogates
-          | baybe/utils
```
Comment (AVHopp) [inline mypy.ini:12]: Is there any specific reason for this chance other than making it more explicit and easier to have more atomic PRs in the future?

File: baybe/targets/numerical.py [hunk 5]
```diff
@@ -92,9 +92,9 @@ def _validate_bounds(self, _: Any, value: Interval) -> None:  # noqa: DOC101, DO
         """
         # IMPROVE: We could also include half-way bounds, which however don't work
         # for the desirability approach
-        if not (value.is_finite or not value.is_bounded):
+        if bounds.is_half_open:
             raise ValueError("Bounds must either be finite or infinite on *both* ends.")
```
Comment (AVHopp) [inline baybe/targets/numerical.py:96]: I would also rephrase this to "Bounds are not allowed to be half-open" or similar as the current message is a bit hard to understand.

**Review by AVHopp — CHANGES_REQUESTED** [review]
We need to discuss and adjust the terminology that we use here as it is just not correct.

File: baybe/utils/interval.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/utils/interval.py:61]: OMG, thanks for pointing out!! As a true German would say: _"Before all the trees, I did not see the wood anymore! ... oder so ähnlich"_ 😄 Was so focused on replacing the previous ambiguous names that I did not pay attention here and really forgot what an open interval is 😭  So my clear suggestion would be: let's use the precise mathematical naming for our properties, no matter if this now requires another refactoring. I can take care of preparing the draft and will ping you once done. Does this sound right?

Comment (AVHopp) [reply baybe/utils/interval.py:61]: Works for me 😄

File: mypy.ini (hunk 4 above)
Comment (AdrianSosic) [reply mypy.ini:12]: I think you missed the actual important change: I wanted to include `interval.py`, that's why I needed to remove the folder ignore and replaced it with file ignores for all contained files **except** `interval.py`

File: baybe/utils/interval.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/utils/interval.py]: So, how about now? Sticking purely to the mathematical definitions of boundedness, hopefully =)

File: baybe/utils/interval.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/utils/interval.py]: Should be fixed now. Please close if you agree

File: baybe/targets/numerical.py (hunk 5 above)
Comment (AdrianSosic) [reply baybe/targets/numerical.py]: "half-open" was wrong to start with, due to my own confusion with the nomenclature. Please see if the new message makes sense and resolve if OK

File: baybe/utils/interval.py (hunk 3 above)
Comment (AdrianSosic) [reply baybe/utils/interval.py]: done, also for the tensor. Looking correct in the rendered doc. Resolve if OK 👍🏼

File: baybe/utils/interval.py (hunk 1 above)
Comment (AVHopp) [reply baybe/utils/interval.py]: Gucci now :)

File: baybe/utils/interval.py (hunk 3 above)
Comment (AVHopp) [reply baybe/utils/interval.py]: Looks good to me. So it does also create the link to the corresponding classes, although they are external references? That is interesting, I did not know that this actually works.

**Review by AVHopp — APPROVED** [review]
Looking gucci to me now :)

File: baybe/utils/interval.py [hunk 6]
```diff
@@ -41,40 +45,60 @@ class Interval:
...
+    def is_right_bounded(self) -> bool:
+        """Check if the interval is right-bounded."""
+        return np.isfinite(self.upper)
+
+    @property
+    def is_half_bounded(self) -> bool:
+        """Check if the interval is half-bounded."""
+        return self.is_left_bounded ^ self.is_right_bounded
+
+    @property
```
Comment (Scienfitz) [inline baybe/utils/interval.py:80]: we really need this deprecation? imo not. if so, it should list in changelog under deprecatoins too

File: baybe/utils/interval.py [hunk 7]
```diff
@@ -41,40 +45,60 @@ class Interval:
...
+        )
+        return np.isfinite(self.lower) and np.isfinite(self.upper)
 
     @property
-    def center(self):
-        """The center of the interval. Only applicable for finite intervals."""
-        if not self.is_finite:
+    def center(self) -> float:
+        """The center of the interval. Only applicable for bounded intervals."""
+        if not self.is_bounded:
```
Comment (Scienfitz) [inline baybe/utils/interval.py:93]: doesnt it make more sense to return `None` rather than throwing an exception?

Comment (AdrianSosic) [reply baybe/utils/interval.py:93]: For me, a relatively clear "no". Because: doing so would change the return type, and this is a common antipattern for good reason: https://docs.quantifiedcode.com/python-anti-patterns/maintainability/returning_more_than_one_variable_type_from_function_call.html?highlight=exception
Basically, one reason to have a custom exception system is to avoid exactly that. 

So I will close this thread, but if you are unhappy, feel free to reopen for further discussion 🙃

File: baybe/utils/interval.py (hunk 6 above)
Comment (AdrianSosic) [reply baybe/utils/interval.py:80]: "Need" is perhaps to strong of a statement 😄 But the same is true for the "BayBE" object before the point of open sourcing, and certain other deprecations. There, we said "let's start using a proper deprecation protocol from now on" and that's what I followed. Having it doesn't hurt and the code is already there, hence I'd say let's leave it and thus I resolve. But if you are really against, I'm not blocking. Feel free to shoot a commit that removes it :D Apart from that, the PR is ready from my end, so let me know if anything else needs change 👍🏼 I'll already prepare the update with the current main.

File: baybe/utils/interval.py (hunk 7 above)
Comment (Scienfitz) [reply baybe/utils/interval.py:93]: This is getting to the crux. The page speaks about should raise an error if a problem is encountered. Imo we don't have a situation like that here, we have a property that can or not exist, neither situations are `errors`. If my user asks my interval whats its center and it has no center, its quite literally telling my user "I have no center" aka `None`. Your variant is also telling the user that, but with a hammer to the face (aka exception).

My variant would enable this pattern
```python
if center := interv.center():
   # beep
else:
   # boop
```
while the existing variant requires:
```python
try:
   center = interv.center()
   # beep
except InfiniteIntervalError:
   # boop
```

Not only is the first variant more `gucci` to the eye, but theres also no need to keep track, import and remember what type of exception this would raise. 

And lastly, it is not recommended to put if-else situations into except statements as they will be slow in the except part. Hence I would perhaps even argue that that is the antipattern we should avoid here (in fact we have several hacky code parts where we use `except` blocks to treat if-else kind of situations for non-error situations).

Comment (AdrianSosic) [reply baybe/utils/interval.py:93]: Hi @Scienfitz
1. Appreciate sharing your thoughts
2. Like the comparison with the hammer 😄 

Upfront: we can also quickly talk about it in the next dev. Nonetheless, here what I think about your answers:
* Yes, I agree that your pattern is more `gucci` to the eye and requires less work, and yes, detailed exception handling is the price you pay for exactness, but I could argue you don't compare against the right situation to start with (and btw: the patterns is also wrong, see end of post ;D). Instead of querying the value (and being hit with the hammer), you could just as well query for the precondition that needs to hold, and in the code logic of the calling code, that  could even be the natural thing to do:
  ```python
  if interval.is_bounded:
     # custom logic involving the use of interval.center
  else:
     # boop
  ```

* Still, you can safely ignore the previous point if you want. The much bigger concern is that in your variant, it suddenly becomes the **responsibility of the caller** to **always** assert that interval `is not None` before it can apply any logic to the returned function value. This becomes necessary for two reasons
  * If you did not do it, your program would no longer be type correct. Look at the actual code where we use it: `lookup[target.name] - target.bounds.center`
    Without the `if interval.center is not None` check upfront, this line could potentially evaluate to "some number minus None". All linters will rightfully complain!
  * Even if you ignored the linters, you will get completely useless error messages like `TypeError: unsupported operand type(s) for -: 'int' and 'NoneType'` in the worst case.
 
Lastly: Note that your pattern is actually incorrect, that is, you really HAVE TO write `if interval.center is not None` instead of `if center := interv.center()` because your expression would evaluate to `False` if the interval happens to be 0.0 (which is a rather common case!)

Let's continue in the meeting :D

Comment (Scienfitz) [reply baybe/utils/interval.py:93]: 1) of course it's the responsibility of the caller to distinguish `None` or `float`. After all, these are both perfectly valid returns of `inverval.center` in my view. **I think this is the core of the disagreement**. If I ask an interval for its center and is has no center, there is neither a fault of the caller nor the interval. It simply has no center and thats valid. No error has been encountered and so no exception should be raised.
If I ask a human on the street `how is your brother?` and they have no brother, what is your preferred answer: i) human says `I have no brother.` or is it ii) human rolls on the floor and yells `THIS IS AN INVALID QUESTION. YOU SHOULD HAVE PERFORMED A BACKGROUND CHECK FIRST`.

2) Ok my pattern suggestion was bugged but quite simply fixed by adding `is not None`. It sill looks better than the alternatives. The variant with `is_bounded` works in both our variants, however i) you dont have any way of enforcing the user to use this instead of center. Implicitly assuming that a user has to check `is_bounded` before they try and access `.center` is really really bad and ii) theres no clear name connection between `is_bounded` and `center` and you can only ensure that a reader gets it by adding an explicit comment -> less optimal. 
Instead `if center:=interv.center is not None:` or `if interv.center is not None:` quite literally reads humanlike `if the interval has a center:`

3) **Wouldn't be a linter complaining about incorrect type in my variant is actually the best outcome?** as it enforced the coder to deal with all possible situations (instead of not dealing with the problem and declaring the solution be a hammer with a nicely engraved human-readable message is thrown at the users face)

Comment (AdrianSosic) [reply baybe/utils/interval.py:93]: @Scienfitz: Changed as agreed in the meeting. Have a look and resolve if OK (+ accept PR if there are no other issues).

Comment (Scienfitz) [reply baybe/utils/interval.py:93]: the fix in `simulation` is not here, assume you wanted to change it in the mypy overhaul of `simulation`?

**Review by Scienfitz — APPROVED** [review]

File: baybe/utils/interval.py (hunk 7 above)
Comment (AdrianSosic) [reply baybe/utils/interval.py:93]: yes, exactly. right now, no fix needed since `simulation` is not yet type checked
